File: packages/agent/src/llm/providers/base/tool-parsing.ts
/**
 * @fileoverview Tool Call Parsing Utilities
 *
 * Provides consistent JSON parsing for tool call arguments across providers.
 * Handles malformed JSON gracefully with logging.
 */

import { createLogger } from '@infrastructure/logging/index.js';

const logger = createLogger('tool-parsing');

// =============================================================================
// Types
// =============================================================================

export interface ToolCallContext {
  /** Tool call ID for logging context */
  toolCallId?: string;
  /** Tool name for logging context */
  toolName?: string;
  /** Provider name for logging context */
  provider?: string;
}

// =============================================================================
// Tool Call Parsing
// =============================================================================

/**
 * Safely parse tool call arguments JSON.
 *
 * This utility provides consistent parsing behavior across all providers:
 * - Handles empty/null arguments by returning empty object
 * - Logs warnings for malformed JSON
 * - Returns empty object on parse failure (fail-open)
 *
 * @param args - Raw JSON string of tool arguments
 * @param context - Optional context for logging
 * @returns Parsed arguments object, or empty object on failure
 *
 * @example
 * ```typescript
 * const args = parseToolCallArguments('{"file": "test.ts"}', {
 *   toolCallId: 'call_123',
 *   toolName: 'read_file',
 *   provider: 'openai',
 * });
 * ```
 */
export function parseToolCallArguments(
  args: string | undefined | null,
  context?: ToolCallContext
): Record<string, unknown> {
  // Handle empty/null args
  if (!args || args.trim() === '') {
    return {};
  }

  try {
    const parsed = JSON.parse(args);
    // Ensure we return an object
    if (typeof parsed === 'object' && parsed !== null && !Array.isArray(parsed)) {
      return parsed;
    }
    // If parsed to non-object, return as-is in wrapper
    logger.warn('Tool arguments parsed to non-object', {
      ...context,
      type: typeof parsed,
      isArray: Array.isArray(parsed),
    });
    return { value: parsed };
  } catch (e) {
    logger.warn('Failed to parse tool call arguments', {
      ...context,
      args: args.length > 200 ? args.slice(0, 200) + '...' : args,
      error: e instanceof Error ? e.message : String(e),
    });
    return {};
  }
}

/**
 * Check if tool call arguments are valid JSON.
 *
 * Useful for validation before processing.
 *
 * @param args - Raw JSON string to validate
 * @returns True if args is valid JSON object
 */
export function isValidToolCallArguments(args: string | undefined | null): boolean {
  if (!args || args.trim() === '') {
    return true; // Empty args are valid (means no arguments)
  }

  try {
    const parsed = JSON.parse(args);
    return typeof parsed === 'object' && parsed !== null && !Array.isArray(parsed);
  } catch {
    return false;
  }
}

File: packages/agent-rs/crates/tron-llm/src/tool_parsing.rs
//! # Tool Call Argument Parsing
//!
//! Safe JSON parsing for tool call arguments received from LLM providers.
//! Handles malformed JSON gracefully — returns empty object rather than erroring,
//! since incomplete tool calls are common during streaming.

use serde_json::{Map, Value};
use tracing::warn;

/// Context for logging when tool call parsing fails.
#[derive(Clone, Debug, Default)]
pub struct ToolCallContext {
    /// The tool call ID (for correlation).
    pub tool_call_id: Option<String>,
    /// The tool name.
    pub tool_name: Option<String>,
    /// The provider that generated this tool call.
    pub provider: Option<String>,
}

/// Parse tool call arguments JSON string into a `Map`.
///
/// Fails open: returns an empty map on parse failure rather than propagating
/// errors, since the agent can still attempt to execute the tool with no args.
///
/// # Arguments
/// * `args` - Raw JSON string from the provider (may be empty, null, or malformed)
/// * `context` - Optional context for warning logs on parse failure
pub fn parse_tool_call_arguments(
    args: Option<&str>,
    context: Option<&ToolCallContext>,
) -> Map<String, Value> {
    let Some(args) = args else {
        return Map::new();
    };

    let trimmed = args.trim();
    if trimmed.is_empty() {
        return Map::new();
    }

    match serde_json::from_str::<Value>(trimmed) {
        Ok(Value::Object(map)) => map,
        Ok(other) => {
            warn!(
                tool_call_id = context.and_then(|c| c.tool_call_id.as_deref()),
                tool_name = context.and_then(|c| c.tool_name.as_deref()),
                provider = context.and_then(|c| c.provider.as_deref()),
                parsed_type = other.to_string().chars().take(20).collect::<String>(),
                "Tool call arguments parsed as non-object, wrapping"
            );
            Map::new()
        }
        Err(e) => {
            warn!(
                tool_call_id = context.and_then(|c| c.tool_call_id.as_deref()),
                tool_name = context.and_then(|c| c.tool_name.as_deref()),
                provider = context.and_then(|c| c.provider.as_deref()),
                error = %e,
                args_preview = &trimmed[..trimmed.len().min(100)],
                "Failed to parse tool call arguments, returning empty object"
            );
            Map::new()
        }
    }
}

/// Validate that a string is valid tool call arguments JSON.
///
/// Returns `true` if the string is valid JSON that parses to an object,
/// or if the string is empty/null (treated as valid empty args).
pub fn is_valid_tool_call_arguments(args: Option<&str>) -> bool {
    let Some(args) = args else {
        return true;
    };

    let trimmed = args.trim();
    if trimmed.is_empty() {
        return true;
    }

    matches!(serde_json::from_str::<Value>(trimmed), Ok(Value::Object(_)))
}

// ─────────────────────────────────────────────────────────────────────────────
// Tests
// ─────────────────────────────────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parse_valid_json_object() {
        let result = parse_tool_call_arguments(Some(r#"{"file": "test.rs"}"#), None);
        assert_eq!(result.len(), 1);
        assert_eq!(result["file"], "test.rs");
    }

    #[test]
    fn parse_empty_object() {
        let result = parse_tool_call_arguments(Some("{}"), None);
        assert!(result.is_empty());
    }

    #[test]
    fn parse_none_returns_empty() {
        let result = parse_tool_call_arguments(None, None);
        assert!(result.is_empty());
    }

    #[test]
    fn parse_empty_string_returns_empty() {
        let result = parse_tool_call_arguments(Some(""), None);
        assert!(result.is_empty());
    }

    #[test]
    fn parse_whitespace_returns_empty() {
        let result = parse_tool_call_arguments(Some("  \n  "), None);
        assert!(result.is_empty());
    }

    #[test]
    fn parse_invalid_json_returns_empty() {
        let result = parse_tool_call_arguments(Some("not json"), None);
        assert!(result.is_empty());
    }

    #[test]
    fn parse_non_object_json_returns_empty() {
        let result = parse_tool_call_arguments(Some("[1,2,3]"), None);
        assert!(result.is_empty());
    }

    #[test]
    fn parse_string_json_returns_empty() {
        let result = parse_tool_call_arguments(Some("\"just a string\""), None);
        assert!(result.is_empty());
    }

    #[test]
    fn parse_complex_object() {
        let args = r#"{"command": "ls -la", "timeout": 5000, "cwd": "/home"}"#;
        let result = parse_tool_call_arguments(Some(args), None);
        assert_eq!(result.len(), 3);
        assert_eq!(result["command"], "ls -la");
        assert_eq!(result["timeout"], 5000);
    }

    #[test]
    fn parse_with_context_logs() {
        let ctx = ToolCallContext {
            tool_call_id: Some("toolu_123".into()),
            tool_name: Some("Bash".into()),
            provider: Some("anthropic".into()),
        };
        // Invalid JSON with context — should still return empty
        let result = parse_tool_call_arguments(Some("broken{"), Some(&ctx));
        assert!(result.is_empty());
    }

    #[test]
    fn parse_nested_objects() {
        let args = r#"{"outer": {"inner": "value"}}"#;
        let result = parse_tool_call_arguments(Some(args), None);
        assert_eq!(result.len(), 1);
        assert!(result["outer"].is_object());
    }

    #[test]
    fn validate_valid_object() {
        assert!(is_valid_tool_call_arguments(Some(r#"{"a": 1}"#)));
    }

    #[test]
    fn validate_empty_is_valid() {
        assert!(is_valid_tool_call_arguments(None));
        assert!(is_valid_tool_call_arguments(Some("")));
    }

    #[test]
    fn validate_non_object_is_invalid() {
        assert!(!is_valid_tool_call_arguments(Some("[1,2]")));
        assert!(!is_valid_tool_call_arguments(Some("\"string\"")));
    }

    #[test]
    fn validate_invalid_json_is_invalid() {
        assert!(!is_valid_tool_call_arguments(Some("not json")));
    }
}
